Moved temp folder form node_modules to the root of project

AleksandrHorev opened the pull request:
The theoretical idea:
It's not a good way to do something with _node_modules_ during the running of tests

A practical problem:
I use a docker container for running my tests on CI. There I have a non-privileged user. By default, _node_modules_ creates as a *read-only* folder and as a result I can't copy from there anything and I get an error like: 
  ```
[Error: EPERM: operation not permitted, copyfile ‘…../node_modules/jest-html-reporters/temp/images/16065009038740.327429410111838.jpg' -> ‘……./html-report/jest-html-reporters-attach/16065009038740.327429410111838.jpg'] {
  errno: -1,
  code: 'EPERM',
  syscall: 'copyfile',
  path: ‘……/node_modules/jest-html-reporters/temp/images/16065009038740.327429410111838.jpg',
  dest: ‘…./html-report/jest-html-reporters-attach/16065009038740.327429410111838.jpg'
}
```

The solution:
Create a temp folder on the root level of the project and remove it after passing of tests.
Additional bonus:
Users can easily get and store all screenshots if needed.

Thank you for your time!
p.s. I open to discuss this problem and finding the best solution
p.s.s. If I set my repo as source of code I'll not have a _dist_ folder to `node_modules/jest-html-reporters/**dist**` by some reasons, so I didn't have a chance to test it from start to end on my CI environment.

AleksandrHorev commented:
Hi @Hazyzh
Could you please review this PR.
Thank you

Hazyzh commented:
> Hi @Hazyzh
> Could you please review this PR.
> Thank you

Sorry for late reply. 

Actually, creating a temp folder to save test result info is a bit hack way whether you do this operation under the `node_modules` folder or not, the reason why we did it is that we can not get any attach information from the Jest native reporter result.

Regarding the issue you met, I think your solution is ok, but it's only good for your specific scenario, we should not change this solution to default. I think creating a `temp` folder on the root level of the project will cause many issues, in fact, we are hard to know which dir is the root level, it's not friendly for many users. 

My suggestion is to add a new reporter option such as `tempDirPath`, the default path is the same as before, and users can also change this temp folder's path in Jest reporter config.

AleksandrHorev commented:
@Hazyzh 
Thank you for your feedback. I updated the PR like you suggested, but I made in a bit another way. `temp path` need only for **addMsg** and for **addAttach**, not for the whole repo, so I extended the _context_ param to object and passed several options there instead of passing it through reporter options. Additionally, it's a bit hard by architecture things to pass these options to _addAttach_ and _addMsg_  from index.js

Hazyzh commented on helper.js:
```diff
@@ -1,25 +1,42 @@
 const addAttach = async (attach, description, context) => {
-  const { testPath, testName } = getJestGlobalData(context);
+  await setUpPaths(context.tempDirPath);
+  const { testPath, testName } = getJestGlobalData(context.context);
```
The param `context` is used here  https://github.com/Hazyzh/jest-html-reporters/pull/84, I am afraid the change here will break the original feature.

AleksandrHorev replied:
Well, on the one hand, you are right.
On the other hand, I'm not a big fan of making methods with 3+ arguments, so I think it's acceptable to reimplement it now and pass all optional params inside this object. It provides an opportunity to extend it in the future if it will be needed.

Hazyzh replied:
sure, you can change the implementation make the param as an object. 

Also, for this specific scenario, how about using the environment variable? 
something like this 
`const dataDirPath = path.resolve(process.env[JEST_HTML_REPORTERS_TEMP_DIR_PATH] || __dirname, "./data");`

AleksandrHorev replied:
yeah, I think for this scenario it will be fine.
in this case, I'll add `JEST_HTML_REPORTERS_TEMP_DIR_PATH ` option to config and make `context` param as object

Hazyzh commented on helper.js:
```diff
@@ -1,25 +1,42 @@
  */
 const addAttach = async (attach, description, context) => {
-  const { testPath, testName } = getJestGlobalData(context);
+  await setUpPaths(context.tempDirPath);
```
If the `setUpPaths` function removes the `tempDir` totally every time, we can only get one piece of data finally.

AleksandrHorev replied:
thank you, I'll think about how to rewrite this point

Hazyzh reviewed (commented):
`Additionally, it's a bit hard by architecture things to pass these options to addAttach and addMsg from index.js`.

indeed agree with this. 

@AleksandrHorev Thanks for your contribution, I have added comments for this `PR`, I will also try to think out a way to resolve this issue.

AleksandrHorev commented:
Hi @Hazyzh 
I updated the PR with your suggestions, please take a look.
Thank you!

Hazyzh reviewed (commented):
> Hi @Hazyzh
> I updated the PR with your suggestions, please take a look.
> Thank you!

Hi @AleksandrHorev , I would like to know how would you use it after this change.
is it like this?
```javascript
const { addAttach, setUpTempDir } = require("jest-html-reporters/helper");
setUpTempDir('custom/path')
test('xxxx', async () => {
await addAttach('* * *')
})

```

AleksandrHorev commented:
> > Hi @Hazyzh
> > I updated the PR with your suggestions, please take a look.
> > Thank you!
> 
> Hi @AleksandrHorev , I would like to know how would you use it after this change.
> is it like this?
> 
> ```js
> const { addAttach, setUpTempDir } = require("jest-html-reporters/helper");
> setUpTempDir('custom/path')
> test('xxxx', async () => {
> await addAttach('* * *')
> })
> ```
Hi @Hazyzh . No, you haven't use setUpTempDir directly. User can set as a reporter parameter
```
['jest-html-reporters', {
      publicPath: './html-report',
      filename: 'report.html',
      expand: true,
     tempDirPath: 'custom/path'
    }],
 ``` 
 or through an environment variable.  `JEST_HTML_REPORTERS_TEMP_DIR_PATH`
 I added to the readme a section about it.

Hazyzh commented:
> > > Hi @Hazyzh
> > > I updated the PR with your suggestions, please take a look.
> > > Thank you!
> > 
> > 
> > Hi @AleksandrHorev , I would like to know how would you use it after this change.
> > is it like this?
> > ```js
> > const { addAttach, setUpTempDir } = require("jest-html-reporters/helper");
> > setUpTempDir('custom/path')
> > test('xxxx', async () => {
> > await addAttach('* * *')
> > })
> > ```
> 
> Hi @Hazyzh . No, you haven't use setUpTempDir directly. User can set as a reporter parameter
> 
> ```
> ['jest-html-reporters', {
>       publicPath: './html-report',
>       filename: 'report.html',
>       expand: true,
>       tempDirPath: 'custom/path'
>     }],
> ```
> 
> or through an environment variable `JEST_HTML_REPORTERS_TEMP_DIR_PATH`
> I added to the readme a section about it.

Have you already pushed all your commits? from current logic, seems it can't work.

AleksandrHorev commented:
> > > > Hi @Hazyzh
> > > > I updated the PR with your suggestions, please take a look.
> > > > Thank you!
> > > 
> > > 
> > > Hi @AleksandrHorev , I would like to know how would you use it after this change.
> > > is it like this?
> > > ```js
> > > const { addAttach, setUpTempDir } = require("jest-html-reporters/helper");
> > > setUpTempDir('custom/path')
> > > test('xxxx', async () => {
> > > await addAttach('* * *')
> > > })
> > > ```
> > 
> > 
> > Hi @Hazyzh . No, you haven't use setUpTempDir directly. User can set as a reporter parameter
> > ```
> > ['jest-html-reporters', {
> >       publicPath: './html-report',
> >       filename: 'report.html',
> >       expand: true,
> >       tempDirPath: 'custom/path'
> >     }],
> > ```
> > 
> > 
> > or through an environment variable `JEST_HTML_REPORTERS_TEMP_DIR_PATH`
> > I added to the readme a section about it.
> 
> Have you already pushed all your commits? from current logic, seems it can't work.

Hm, why do you think so? I tested it on my repo and it worked fine (at least with the param in config)

AleksandrHorev commented:
@Hazyzh Thank you a lot! You were right, I made a mistake, fixed
```
init() {
    removeTempDir(this._options.tempDirPath);
    setUpTempDir();
    this.initAttachDir();
  }
  ```

Hazyzh commented:
> @Hazyzh Thank you a lot! You were right, I made a mistake, fixed
> 
> ```
> init() {
>     removeTempDir(this._options.tempDirPath);
>     setUpTempDir();
>     this.initAttachDir();
>   }
> ```

![image](https://user-images.githubusercontent.com/21355783/102611278-97dac300-4169-11eb-844d-bb1c52b109fd.png)

AleksandrHorev commented:
> > @Hazyzh Thank you a lot! You were right, I made a mistake, fixed
> > ```
> > init() {
> >     removeTempDir(this._options.tempDirPath);
> >     setUpTempDir();
> >     this.initAttachDir();
> >   }
> > ```
> 
> ![image](https://user-images.githubusercontent.com/21355783/102611278-97dac300-4169-11eb-844d-bb1c52b109fd.png)

![image](https://user-images.githubusercontent.com/18176026/102611627-541f9380-4140-11eb-9aea-297a16bdf972.png)
there a private function **_setTempDiPathVariable() ** that stores this path. You need to pass in once and it will work. But now I see it's unnecessary overengineering, I'll make it easier in 5 minutes.

Hazyzh commented:
@AleksandrHorev haha, Don't worry, there is no time limit😄.
 
Generally adds a new config is an elegant solution, but I am afraid that it can not work here.

It may work fine when you exec Jest with only one test file. But it may fail when you run Jest with multiple test files. that's because the Just will create multiple child processes to run those specific tests when you have multiple test files. 

This means that the `index.js` was running on the main process and you exec the `setUpTempDir` function also in the main process. But for a specific test file,  the code was running in a child process, so it can not get the correct value of the variable `attachDirPath ` `tempDirPath `. You can add `console.log` in `addAttach` function and have a test.

Hazyzh commented:
And we can get the environment variable in both the main process and all child processes, so here I think we can only support environment variable configuration for this feature.

AleksandrHorev commented:
hm, got your idea. But I think `[TEMP_DIR_PATH]: __dirname,` as a default value will work in all processes and should be fine

AleksandrHorev commented:
> And we can get the environment variable in both the main process and all child processes, so here I think we can only support environment variable configuration for this feature.

I tested the current solution, you were right, there some problems with the context. I'll reimplement, do more tests and update the PR.

Hazyzh commented:
> > And we can get the environment variable in both the main process and all child processes, so here I think we can only support environment variable configuration for this feature.
> 
> I tested the current solution, you were right, there some problems with the context. I'll reimplement, do more tests and update the PR.

hi @AleksandrHorev, I push a pr to your branch https://github.com/AleksandrHorev/jest-html-reporters/pull/2/files, I think the environment variable support is enough for this scenario. And we needn't modify much logic here. Please help review.

AleksandrHorev commented:
Hi @Hazyzh, I approved and merged PR. 
Thank you for your help!

Hazyzh commented:
Hi @AleksandrHorev, I have published a new version (v2.1.2) with this feature. you can try to use it. 

Thanks for your contribution again!
